### PR TITLE
Fixed 29336 -- Added docs for circular template inheritance.

### PR DESCRIPTION
**Pull request by Adrian-Turjak**
Deals with: https://code.djangoproject.com/ticket/29336

### CODE REVIEW

File: docs/ref/templates/language.txt [hunk 1]
```diff
@@ -434,6 +434,67 @@ wouldn't know which one of the blocks' content to use.
...
+
+Consider the following project:
+
+.. snippet::
+    :filename: settings.py
+
+    INSTALLED_APPS = (
+        "app1",
+        "app2",
+        "app3",
```
Comment (carltongibson) [inline docs/ref/templates/language.txt]: I think these need to be the reversed (or the templates adjusted). As it stands `app1/templates/page.html` will be found first. This extends nothing so will just be rendered as is.

**Comment by Adrian-Turjak** [comment]
@carltongibson Thanks for the review! Will adjust the patch, and move stuff to the overriding templates section. I think I will still leave a small example in the current place because this is technically another form of template inheritance and we should promote it. Then on the overriding page I'll add a more thorough example, and maybe even rewrite part of that page to encourage overextension over outright overriding. Does that sound sane?

**Comment by Adrian-Turjak** [comment]
Also, designers using this feature do need to vaguely understand how the loading works at least a little, so a short note is useful that explains very briefly what is talked about here:
https://docs.djangoproject.com/en/dev/ref/templates/api/#loader-types

With a link to that reference if they want to dig further.

**Comment by Adrian-Turjak** [comment]
@carltongibson Alright, I've done a bit of a rewrite. I've kept the main bulk of the change in `language.txt` because this is a language feature that feels better documented there than in a how-to guide that isn't outright related (the guide is about overriding, not template inheritance).

I've updated the how-to guide to reflect that overextension is in many ways the better alternative to outright overriding, and tried to make it more about how they both work rather than entirely giving examples.

I think this now mostly makes sense and will hopefully guide people to actually make use of this feature smartly.

File: docs/howto/overriding-templates.txt [hunk 2]
```diff
@@ -1,26 +1,46 @@
-====================
-Overriding templates
-====================
+=====================================
+Overriding or Overextending templates
```
[the remark this answers is not in the document]
Comment (Adrian-Turjak) [reply docs/howto/overriding-templates.txt]: Cool, can agree with that.

File: docs/ref/templates/language.txt [hunk 3]
```diff
@@ -431,6 +431,57 @@ it also defines the content that fills the hole in the *parent*. If there were
 two similarly-named :ttag:`block` tags in a template, that template's parent
 wouldn't know which one of the blocks' content to use.
 
+.. _circular-template-inheritance:
+
+Circular Template inheritance
```
[the remark this answers is not in the document]
Comment (Adrian-Turjak) [reply docs/ref/templates/language.txt]: Will do.

**Comment by Adrian-Turjak** [comment]
@carltongibson Yeah I wasn't sure what term to go with since Recursive Inheritance seemed like a mouthful, but more in line with Django than overextension. Recursive inheritance it is!

Alright, will take the changes a step back and take into account your comments. :)

Thanks for your feedback!

**Comment by Adrian-Turjak** [comment]
@carltongibson Alright, another attempt!

I've decided to keep the explanation and example in `language.txt` because on reflection it made far better sense in the same context as inheritance than overriding. This isn't overriding, it's just another aspect of Django's template inheritance and keeping the focus on it as a language feature rather than a note in the how-to feels cleaner. 

I've added links between the how-to and the new recursive inheritance doc where I think it feels best related, but keeping the focus on: "this is how you use another part of the language" vs "this is how you use the language to override things" seems a better approach.

**Comment by Adrian-Turjak** [comment]
We also probably want to consider what the verb action for this is, and if I should edit the doc to include a use of it. Saying "override this template" is contextually wrong and implies something else, so  maybe "recursively extend this template" may be more appropriate.

I only bring this up since I'm trying to make the docs I'm writing elsewhere consistent with what we might want here in the Django docs.

My docs elsewhere:
```
Additionally, some scripts require you to place them within the page's <head>
tag. To do this, recursively extend the ``base.html`` templates in your theme
to override the ``custom_head_js`` block.
```

If we use a certain terminology here, others might use it too and things may be a bit more consistent when people try to explain a feature or find the right docs.

**Comment by rfleschenberg** [comment]
A template named `base.html` without an additional namespace is commonly defined at the project level, as the main template for the whole project. I suggest using a different name for this example. Maybe `reuseable_app/reuseable_app_base.html` would be better?

**Comment by Adrian-Turjak** [comment]
@rfleschenberg Good point! Will do.

**Comment by carltongibson** [comment]
Hi @Adrian-Turjak. Thanks for the effort here. Alas I still don't agree with the approach.

The re-write is too substantial. A **small** tip in the inheritance section of `language.txt` linking to a fuller discussion _at the end_ of the existing How-To is all that is needed.

**Comment by carltongibson** [comment]
Hi @Adrian-Turjak — Are you still up for working on this issue? (If not I can deassign it and someone else is likely to pick it up...) Thanks.

**Comment by Adrian-Turjak** [comment]
@carltongibson Sorry I entirely totally forgot about this.

I still don't fully agree that this should be thought of as overriding. It is just template inheritance used a little differently, with a mindset change around how you make templates extensible. Especially if you are being clever with your use of `block.super`. The only real overlap it has with overriding is the loading order logic, but beyond that is just a feature of the inheritance logic itself.

My worry is that a linking it heavily with overriding will give people the wrong use case examples, and by keeping it just a tip in `language.txt` will mean this powerful/useful feature gets overlooked. It probably doesn't need to be as big as I've written it, but I'm out of ideas where else to put it that doesn't make this seem like purely an overriding related feature.

In all the cases where I use this (and used the old Overextends library) it was always to **avoid** having to override, and instead extend a template of the same name with a minor incremental change, often multiple times across multiple apps that each optionally add items around existing block content.

File: docs/ref/templates/language.txt [hunk 4]
```diff
@@ -431,6 +431,104 @@ it also defines the content that fills the hole in the *parent*. If there were
...
+    {% block main %}
+    <p>A paragraph in the reusable app.</p>
+    {% endblock %}
+
+    {% block footer %}
+    <footer>Footer text from the reusable app</footer>
+    {% endblock %}
+
+Here we have a reusable app that gives us a ``reuseable_app_base.html``
+template. Let's imagine for this example that this is a useful third-part app,
```
Comment (phildini) [inline docs/ref/templates/language.txt]: `third-party` instead of `third-part`, I think?

File: docs/ref/templates/language.txt [hunk 5]
```diff
@@ -431,6 +431,104 @@ it also defines the content that fills the hole in the *parent*. If there were
...
+and that there are many other apps built to compliment it.
```
Comment (phildini) [inline docs/ref/templates/language.txt]: I think we want `complement` not `compliment` here.

**Comment by carltongibson** [comment]
Hi @phildini Thanks for the review. Gives it a bump 🙂

@Adrian-Turjak: 

> ...that doesn't make this seem like purely an overriding related feature.

That's just my point. It **is* a purely overriding feature. A cool one, yes. 🙂
As such, an extended discussion is best placed in the existing How-To. 
Adding a smaller note the language guide, and linking again to the How-To for the bigger example, should be sufficient. 
Make sense?

Still have energy for this? I think it's a worth while change if you do. 🙂

**Comment by Adrian-Turjak** [comment]
> That's just my point. It *is a purely overriding feature. A cool one, yes.

But it isn't! :disappointed:  

It is using inheritance `extend` syntax, it is able to use `{{ block.super }}` and use the blocks from the template it inherits from. It is inheritance pure and simple. It _can_ be used for overriding, but in the same way that inheritance itself can. That's why I think throwing in with overriding defeats the purpose because it makes people think its all about overriding. The only part it shares with overriding is that the template lookup logic is based on how the override order is found.

I don't use it to override. I use it to make templates that can be recursively extended by having each template add a little extra to the previous one (like inheritance) without having to make lots of differently named templates and manually inheriting them all somehow. I do it because I need 3+ different apps which can add just a little extra to the same view template, but I can't rely on them all inheriting each other's templates to have a chain, I need them to be able to be mixed and matched and still work. This feature gives me that. That's what made the old library Overextends useful and what most people used it for. It just a shame about the name including 'over' in it, but the key part there is 'extends'.

The reason people think of it as overriding is because Django used to treat same name templates as overrides (which is what the how-to doc you want this placed in says), and using `extends` on a same name template would cause [weird things](https://code.djangoproject.com/ticket/22232) because the lookup logic didn't look at previously overridden templates. Instead the fix that allowed templates to extend same name templates uses the full list of know same name templates and inherits them in the order they were loaded. That is inheritance.

This is the feature I am documenting:
https://code.djangoproject.com/ticket/15053
^ overriding is not talked about. It's an inheritance feature.

As a middle ground I'll add just a note to `language.txt`, because I agree that placing it in `language.txt` is probably wrong and this is an advanced feature. Then I'll make a new how-to for recursive-inheritance, and add some notes between it and the overriding how-to.

File: docs/ref/templates/language.txt (hunk 4 above)
Comment (Adrian-Turjak) [reply docs/ref/templates/language.txt]: Thanks! Fixed.

File: docs/ref/templates/language.txt (hunk 5 above)
Comment (Adrian-Turjak) [reply docs/ref/templates/language.txt]: yes!

**Comment by Adrian-Turjak** [comment]
Alright, hows this?

**Comment by Adrian-Turjak** [comment]
Weird test failure... seems to be a jenkins error or something.

@carltongibson Is this newest version better? I'm sorry for being a bit more stubborn that I should be, but I really would like to avoid lumping this feature in with overriding, and I know you're just trying to make sure the Django docs stay sensible. :)

**Comment by carltongibson** [comment]
Hi @Adrian-Turjak. Yes... this crossed my mind the other day... More or less I still think this is just a species of template inheritance, and doesn't merit a separate how-to of it's own. I shall review it though over the coming weeks. Let's get something in. Thanks for your effort. 🙂

**Comment by rfleschenberg** [comment]
Do we want to encourage usage patterns that rely on the order of `INSTALLED_APPS`? Personally I dislike this, but maybe that't just me.

**Comment by carltongibson** [comment]
@rfleschenberg if you’re going to use the App Directories loader then you’re going to have this. Personally I think disabling it is a bit of an advanced optimisation.

**Comment by rfleschenberg** [comment]
@carltongibson I use the app directories loader, but I do not override app templates from other apps. If I want to override an app template, I do so from a separate project-wide template directory that is configured in `TEMPLATES[0]['DIRS']`. That way, the order of `INSTALLED_APPS` does not matter, as long as all apps use proper namespacing.

I thought this was the official recommendation, but I might be wrong (I cannot find anything about this in the current version of the docs).

**Comment by carltongibson** [comment]
Yes, that’s usual I’d say. (Not seeing your point then really)

All this PR will end as is an extra section explaining that you can “name the same” to get the “recursive” effect. I don’t see a change to the existing explanations.

**Comment by rfleschenberg** [comment]
My point was that it might be worth considering not documenting (and thereby to a certain extent encouraging) it, because relying on the order of `INSTALLED_APPS` for such things is not nice, in my experience. But, maybe that's just me. I just wanted to make sure it gets considered :)

**Comment by Adrian-Turjak** [comment]
@rfleschenberg This is a feature of the template language, and one that in very specific cases is VERY useful. When dealing with multiple layers of templates, and plugins, there isn't a good alternative when you don't want to override a whole template and multiple layers each need to add something extra to the same pages.

What you also aren't getting is what exactly this feature does. This isn't overriding, it's inheritance, but specific to the same name template you are inheriting from. Thus you can selectively extend that same template without having to override the whole thing as you may have to do using your existing pattern. In fact, nothing stops you from still having your own templates in a folder and using just one layer of recursive inheritance. But this feature isn't likely to be used by simple projects, rather by many layered ones with plugin support.

This was added to django 1.9, but before then many of us got by using:
https://github.com/stephenmcd/django-overextends

When doing any form of template loading from apps, there is little to nothing that can be avoided in regards to understanding app loading order. In fact Django has many cases (and apps) which rely on app order, with some libraries explicitly saying 'load me first, or before X'. If you are using Django, you are relying on app order, and templates is another layer of that already existing pattern.

**Comment by rfleschenberg** [comment]
@Adrian-Turjak Just because I raise a concern here, that doesn't mean I do not understand this feature. I'd appreciate it if you don't make such assumptions in the future.

I am also fully aware that other things in the Django ecosystem rely on the order of `INSTALLED_APPS`, too. It is part of the reason why I see this as a potential drawback (because you can get conflicts).

As I have written before, I am not strongly opposed to documenting this, I just wanted to make sure this concern is given a thought.

**Comment by Adrian-Turjak** [comment]
@rfleschenberg No slight was meant, purely that until you've made use of this feature in anger, it's not quite all that easy to fully grasp both the usefulness and the benefits compared to overriding. I made the assumption that you haven't made use of this, if I'm wrong, my apologies.

I understand your worry about relying on `INSTALLED_APPS` but I think that's a given, and that the benefits of features like this outweigh the potential worry.

I just want it documented so I can link people to this in some of my docs, but also because it encourages people to override less and extend more even in same name templates and allows building more maintainable apps.

**Comment by Adrian-Turjak** [comment]
I made the example into an actual project:
https://github.com/Adrian-Turjak/recursive_template_test

And I think I do need to find a good place in the new docs to be VERY emphatic that this feature is very heavily reliant on template look up and explain specifically how.

**Comment by frnhr** [comment]
$0.02 about naming... Just trying to set some common language here. I'm assuming all of this below is already well known to everyone reading this thread. But maybe people have different starting points for thinking about things and are therefore using different words for same things. So, here follows some over-communication, in the hope of happiness and progress.

-----

IMHO the feature that we all use and love is **both** _override_ and _extend_. The naming of the `overextends` library / template tag was fantastically spot-on.

1. The template engine will first find the template in `my_fancy_app` (or in the project template folder), *instead* of the template with the same name in the `external_library_app` app. This is *overriding*, as defined [in the docs](https://docs.djangoproject.com/en/2.2/howto/overriding-templates/):
    > ### Overriding templates
    >
    > In your project, you might want to override a template in another Django application, whether it be a third-party application or... 

2. The template engine then proceeds to find the parent template for the loaded template. It correctly finds the parent template inside the `external_library_app`. This is *extending*, as defined [in the docs](https://docs.djangoproject.com/en/2.2/ref/templates/language/#template-inheritance):
    > ### Template inheritance 
    > ...
    > The **extends** tag is the key here. It tells the template engine that this template “**extends**” another template. When the template system evaluates this template, first it locates the parent...

So, it's both.  🗞☕️🙂 

More broadly speaking all of these are possible (and all very useful at times):
 * override without extending
 * extend without overriding
 * override and extend

-----

One thing it is arguably not, is _recursion_. IMHO recursion is the very thing which ticket [15053](https://code.djangoproject.com/ticket/15053) (and the overextends library before it) have been successful at *avoiding*. 

Recursion would happen if a template could somehow extend _itself_ and survive. The criterion for "_itself_" should be the absolute filesystem path of the template file.

**Comment by carltongibson** [comment]
OK, so this has been sat here a while... Reviewing again I still think it's over-complicating matters. 

* The small addition to `language.txt` is about right.
* Then an extra section at the bottom of `overriding-templates.txt` explaining the override and extend usage is all we need. 
   * A example showing, say, adding an `` to the branding block in `admin/base_site.html` or similar would be appropriate. 

I don't want to add a whole other how-to for this. Yes, let's call it out. But that's it.

**Comment by felixxm** [comment]
Closing due to inactivity.

**Pull request closed without merging**